src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp

```diff
@@ -432,187 +867,194 @@ class npc_memory : public CreatureScript
-                        case 1:
-                            AddWaypoint(0, 746.73f, 650.24f, 411.56f);
-                            break;
-                        case 2:
-                            AddWaypoint(0, 781.32f, 610.54f, 411.82f);
-                            break;
-                    }
-                    break;
-                case NPC_PRIESTESS:
-                    switch (uiType)
-                    {
-                        case 0:
-                            AddWaypoint(0, 715.06f, 637.07f, 411.91f);
-                            break;
-                        case 1:
-                            AddWaypoint(0, 750.72f, 650.20f, 411.77f);
-                            break;
-                        case 2:
-                            AddWaypoint(0, 779.77f, 607.03f, 411.81f);
-                            break;
-                    }
-                    break;
-            }
-
-            Start(false, true);
-            uiWaypoint = uiType;
+            BeforeHit += SpellHitFn(spell_eadric_hammer_of_righteous_SpellScript::HandleDamage);
         }
+    };
+
+    SpellScript* GetSpellScript() const override
+    {
+        return new spell_eadric_hammer_of_righteous_SpellScript();
+    }
+};
+
+class spell_eadric_hammer_of_righteous_faceroller : public SpellScriptLoader
+{
+public:
+    spell_eadric_hammer_of_righteous_faceroller() : SpellScriptLoader("spell_eadric_hammer_of_righteous_faceroller") { }
 
-        void UpdateAI(uint32 uiDiff) override
+    class spell_eadric_hammer_of_righteous_faceroller_SpellScript : public SpellScript
+    {
+        PrepareSpellScript(spell_eadric_hammer_of_righteous_faceroller_SpellScript);
+
+        bool Load() override
         {
-            EscortAI::UpdateAI(uiDiff);
+            Unit* caster = GetCaster();
+            return caster->GetTypeId() == TYPEID_PLAYER && caster->GetMap()->IsHeroic();
+        }
 
-            if (!UpdateVictim())
+        void CheckHealth(SpellEffIndex /*effIndex*/)
+        {
+            Creature* creature = GetHitCreature();
+            if (!creature || creature->GetEntry() != NPC_EADRIC)
                 return;
 
-            DoMeleeAttackIfReady();
+            // If the spell's damage is higher than Eadric' health,
+            // player gets the achievement
+            if (GetHitDamage() >= static_cast<int32>(GetHitUnit()->GetHealth()))
+                creature->AI()->SetData(DATA_FACEROLLER_ACHIEVEMENT, 1);
         }
 
-        void JustDied(Unit* /*killer*/) override
+        void Register() override
         {
-            instance->SetData(DATA_ARGENT_SOLDIER_DEFEATED, instance->GetData(DATA_ARGENT_SOLDIER_DEFEATED) + 1);
+            OnEffectHitTarget += SpellEffectFn(spell_eadric_hammer_of_righteous_faceroller_SpellScript::CheckHealth, EFFECT_0, SPELL_EFFECT_SCHOOL_DAMAGE);
         }
     };
 
-    CreatureAI* GetAI(Creature* creature) const override
+    SpellScript* GetSpellScript() const override
```

Comment:
all of these could use the generic spell script loader (see fb87ac8e8da3d215d4988b8c2fa25d7a98cf9ee0)